### PR TITLE
Add configUSE_SCHEDULER_CORE_MASK and vTaskSetSchedulerCoreMask API

### PR DESCRIPTION
**Pull request by kuopinghsu**
# Add configUSE_SCHEDULER_CORE_MASK and vTaskSetSchedulerCoreMask API

Description
-----------
Adds a new SMP-only feature that lets the application control which
cores the FreeRTOS scheduler is permitted to run non-idle tasks on, at run
time, without stopping the scheduler.

The primary use cases are:

1. **Dynamic load management** -- exclude one or more cores from scheduling
   non-idle work so they can be idled or throttled without stopping the
   scheduler on the remaining cores.

2. **Core power-off preparation** -- before powering off a secondary core
   (cores 1..N-1), clear that core's bit in the scheduler mask so the
   scheduler stops assigning new tasks to it.  Once the core's current task
   yields and the core is running only its idle task, the platform power
   management layer can safely gate the clock or cut power to that core.
   Core 0 must never be masked out because it owns the tick and the
   scheduler entry point.

A new configuration macro `configUSE_SCHEDULER_CORE_MASK` gates the entire
feature.  Setting it to `0` produces a clean preprocessor rollback to the
original code with zero overhead.

**Files changed:**

| File | Change |
|---|---|
| `tasks.c` | +113 / -14 |
| `include/task.h` | +38 / 0 |
| `include/FreeRTOS.h` | +8 / 0 |
| `examples/template_configuration/FreeRTOSConfig.h` | +8 / 0 |

**Details:**

`include/FreeRTOS.h`: added default `#define configUSE_SCHEDULER_CORE_MASK 1`
and a compile-time `#error` that blocks use on single-core builds.

`include/task.h`: added declarations for both new API functions under the
combined `configNUMBER_OF_CORES > 1 && configUSE_SCHEDULER_CORE_MASK == 1` guard.

`tasks.c`: added the `uxSchedulerCoreMask` static variable; added three
`#if ( configUSE_SCHEDULER_CORE_MASK == 1 )` checks inside the SMP task
selection loop so that non-idle tasks are never assigned to or kept on a
masked core; added implementations of both API functions.

`examples/template_configuration/FreeRTOSConfig.h`: added
`configUSE_SCHEDULER_CORE_MASK 1` to the SMP section.

**Opt-out:** set `#define configUSE_SCHEDULER_CORE_MASK 0` in `FreeRTOSConfig.h`.
Every added line is inside a preprocessor guard, so the build is identical
to the unpatched kernel.

**New API** (available when `configNUMBER_OF_CORES > 1` and
`configUSE_SCHEDULER_CORE_MASK == 1`):

```c
/* Bit N = 1 means core N may run non-idle tasks.
 * Bit N = 0 means core N runs only its idle task.
 * Pass ( tskNO_AFFINITY ) to re-enable all cores. */
void vTaskSetSchedulerCoreMask( UBaseType_t uxCoreMask );

/* Returns the current global scheduler core mask. */
UBaseType_t uxTaskGetSchedulerCoreMask( void );
```

Usage example -- power off core 2 on a 4-core system:

Note: masking core 0 is valid at the scheduler level (core 0 will just run
only its idle task). What must not be done is powering off the core that
hosts the tick source (typically core 0), because the kernel tick and
scheduler entry point depend on it. Masking is a safe prerequisite step
before powering off any secondary core.

```c
/* Step 1: stop the scheduler from assigning tasks to core 2. */
UBaseType_t uxMask = uxTaskGetSchedulerCoreMask();
uxMask &= ~( ( UBaseType_t ) 1U << 2U );
vTaskSetSchedulerCoreMask( uxMask );

/* Step 2: wait (platform-specific) until core 2 is running only idle. */
waitForCoreIdle( 2 );

/* Step 3: power off core 2 via the BSP. */
bspCorePowerOff( 2 );

/* To bring it back: power on, then re-enable in the mask. */
bspCorePowerOn( 2 );
vTaskSetSchedulerCoreMask( tskNO_AFFINITY );
```

Test Steps
-----------
1. Build an SMP application with `configNUMBER_OF_CORES > 1` and
   `configUSE_SCHEDULER_CORE_MASK 1` (default).  Confirm it compiles and
   links without warnings.
2. Call `vTaskSetSchedulerCoreMask( 1U )` at run time (only core 0 enabled).
   Confirm that tasks on cores 1..N-1 are no longer scheduled and those
   cores run only their idle tasks.
3. Call `vTaskSetSchedulerCoreMask( tskNO_AFFINITY )`.  Confirm normal
   scheduling resumes on all cores.
4. Build with `configUSE_SCHEDULER_CORE_MASK 0`.  Confirm it compiles
   cleanly and that neither `vTaskSetSchedulerCoreMask` nor
   `uxTaskGetSchedulerCoreMask` is referenced in the output binary.
5. Build with `configNUMBER_OF_CORES 1` and any non-zero value of
   `configUSE_SCHEDULER_CORE_MASK`.  Confirm the expected `#error` fires.

Checklist:
----------
- [x] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: include/task.h [hunk 1]
```diff
@@ -1420,6 +1420,45 @@ BaseType_t xTaskResumeFromISR( TaskHandle_t xTaskToResume ) PRIVILEGED_FUNCTION;
...
 
+#if ( ( configNUMBER_OF_CORES > 1 ) && ( configUSE_SCHEDULER_CORE_MASK == 1 ) )
+
+/**
+ *
+ * Controls which cores are allowed to run non-idle tasks system-wide.
+ * Bit N = 1 means core N may run tasks; bit N = 0 means core N will only
+ * run its idle task.  configNUMBER_OF_CORES must be greater than 1 for this
+ * function to be available.
+ *
```
[the remark this answers is not in the document]
Comment (kuopinghsu) [reply include/task.h:1440]: I think it can mask Core 0 since the scheduler is still running. Note that Core 0 specifically cannot be powered down. This behavior is acceptable; this can skip the Core 0 check.

[the remark this answers is not in the document]
Comment (kuopinghsu) [reply include/task.h:1440]: Update comment. Passing 0 as the mask is valid.

File: tasks.c [hunk 2]
```diff
@@ -3099,6 +3128,62 @@ static void prvInitialiseNewTask( TaskFunction_t pxTaskCode,
 
 /*-----------------------------------------------------------*/
 
+#if ( ( configNUMBER_OF_CORES > 1 ) && ( configUSE_SCHEDULER_CORE_MASK == 1 ) )
+    void vTaskSetSchedulerCoreMask( UBaseType_t uxCoreMask )
+    {
+        BaseType_t xCoreID;
+        UBaseType_t uxOldMask;
```
[the remark this answers is not in the document]
Comment (kuopinghsu) [reply tasks.c:3141]: add traceENTER_vTaskSetSchedulerCoreMask( uxCoreMask ), traceRETURN_vTaskSetSchedulerCoreMask(), traceENTER_uxTaskGetSchedulerCoreMask() and traceRETURN_uxTaskGetSchedulerCoreMask( uxCoreMask ) trace hooks

**Review by kuopinghsu — COMMENTED** [review]
Applied Copilot's recommended changes.

**Comment by chinglee-iot** [comment]
Hi kuopinghsu,

I would like to share my observations regarding this implementation:

1. **Enforcing Context Switches on Mask Activation:** When a core is newly added to the scheduler mask via `vTaskSetSchedulerCoreMask`, the scheduler needs to immediately request a context switch on that specific core. This ensures that the newly enabled core stops idling and promptly starts executing available ready tasks.

2. **Active Idle Task Power-Off:** The following issue occurs on the flip side when powering down a core. My understanding is that when a core is excluded via the scheduler mask, it should safely transition to running the internal idle task while waiting to be physically powered off. However, if this active idle task becomes permanently blocked or frozen by the power-off sequence, the system loses that core's ability to handle essential background maintenance, such as task termination and resource cleanup (`prvCheckTasksWaitingTermination`).

Would you be open to adding functional test cases to the FreeRTOS repo under the SMP CMock directory?
https://github.com/FreeRTOS/FreeRTOS/tree/main/FreeRTOS/Test/CMock/smp

Adding these tests would greatly help the reviewers catch these multi-core edge cases and better understand the functional intent of this feature.

**Comment by kuopinghsu** [comment]
Update newly added to the scheduler mask via `vTaskSetSchedulerCoreMask`, the scheduler yields that specific core

**Comment by chinglee-iot** [comment]
> Would you be open to adding functional test cases to the FreeRTOS repo under the SMP CMock directory? https://github.com/FreeRTOS/FreeRTOS/tree/main/FreeRTOS/Test/CMock/smp
> 
> Adding these tests would greatly help the reviewers catch these multi-core edge cases and better understand the functional intent of this feature.

I've opened a pull request to add these unit tests under the SMP CMock directory:
https://github.com/FreeRTOS/FreeRTOS/pull/1414
Please let me know your thoughts and if any adjustments are needed.
